### ISSUE TITLE
Why does bench_omni.ipynb use raw counts instead of log-transformed expression?

### ISSUE THREAD

**Issue by Jiaxin-Fan527** [cantinilab/scPRINT#55]
Hi, thank you for making this project available.

I have a question regarding the preprocessing in `notebooks/bench_omni.ipynb`.

I noticed that the input expression matrix appears to be raw counts. Looking through the code, I found that the preprocessor is initialized as:

```python
preprocessor = Preprocessor(force_preprocess=True,skip_validate=True, do_postp=False,)
```

My understanding is that `do_postp=False` disables the post-processing step (including log transformation), so the model seems to receive raw counts directly.

Could you clarify whether this is the intended input format? If so, is there any internal log transformation applied later in the pipeline, or was the model trained to operate directly on raw counts?

Thanks for your help.

**Comment by jkobject** [comment]
Hello @Jiaxin-Fan527 

No, postp doesn't apply  log transform. the datasets are never transformed in advance.
in fact do_postp only does transformation to compute neighbors but it stores it outside the anndata.X field.

the transformation happens within the model itself (in model.py) 

Best,

**Comment by Jiaxin-Fan527** [comment]
Thank you for the clarification.

I noticed that the model documentation mentions that the input data should be log-transformed. However, in the gene regulatory tutorial (`notebooks/bench_omni.ipynb`), the following warning appears in Cell 8:

> WARNING: It seems you use rank_genes_groups on the raw count data. Please logarithmize your data before calling rank_genes_groups.

This made me wonder whether there might be a mismatch between the preprocessing used in the tutorial and the preprocessing expected by the model. Am I interpreting this correctly, or is the warning only related to the downstream `rank_genes_groups` analysis rather than the actual model input?

In addition, while looking through the code, I could not find any explicit log-transformation step inside the `GNInfer` class. Therefore, I am not sure whether the input is expected to be pre-log-transformed by the user before being passed to the model.

Could you please clarify where the log transformation is applied in the workflow?

Thank you for your help.

**Comment by jkobject** [comment]
Thanks for following up — your interpretation of the warning is correct, and my previous answer was too broad.

For the `medium-v1.5.ckpt` used by `notebooks/bench_omni.ipynb`, the intended model input is raw counts. The exact path is:

- `GNInfer.predict` builds a `Collator`, then takes `batch["x"]` as `expression` (`scprint/tasks/grn.py`, around lines 207–258). It does not transform `adata.X` itself.
- `GNInfer` passes that tensor to `model._predict`, which calls `model.forward`.
- The expression transform is in `scPrint._encoder` (`scprint/model/model.py`, around lines 545–553): when `model.normalization == "log"`, it computes `torch.log2(1 + expression)` before the expression encoder.
- The notebook checkpoint stores `normalization="log"`; checkpoint loading restores that setting (`model.py`, around line 431). So this specific model sees `log2(1 + raw_counts)` internally. Users should not pre-log-transform the matrix passed to scPRINT. (Other checkpoints can carry a different `normalization` setting, so the checkpoint value is authoritative.)

The Scanpy warning is a separate issue. With `how="most var across"`, `GNInfer.predict` calls `sc.tl.rank_genes_groups` on the supplied `adata` before model inference (`scprint/tasks/grn.py`, around lines 170–179). In this notebook that object still contains raw counts, whereas Scanpy expects log-transformed values for `rank_genes_groups`. Therefore the warning is about the downstream gene-selection step, not about the representation consumed by the scPRINT model. It is a real issue in that notebook/GNInfer path: ranking should be computed on a normalized/log-transformed copy while raw counts are preserved for the model.

I have recorded that as a separate implementation follow-up. Thanks for catching the distinction.

**Comment by Jiaxin-Fan527** [comment]
> sc.tl.rank_genes_groups

Thanks for the detailed clarification and for confirming the distinction. That makes sense now. Thanks again!
